_posts/2015-02-20-elasticsearch-and-denormalization.md

```diff
@@ -0,0 +1,223 @@
+response times can be unacceptably slow. Heroku web dynos time out at thirty
+seconds and if your app is regularly butting up against that limit in the course
+of normal operation then it's probably time to think about some serious
+refactoring.
+
+At Stitch Fix we have such a tool: the Receiving Queue. It is designed to give
+an itemized list of what new inventory we have received on a particular day, to
+show ordered vs received item counts and detailed information about the style
+of clothing received, so that the warehouse team can be sure they have received
+the items we were expecting.
+
+When we launched the Receiving Queue it quickly became apparent that performance
+was a problem. At first we could only load a week's worth of receipts, then a 
+week for a single warehouse, and then only a single day. It worked but since the
+data view was so narrow it was not as powerful as we intended it to be.
+
+A few months later a similar project arose. This time it was the Sample Queue, a
+tool that would allow our Sample Coordinators to see, based on purchase orders,
+what clothing samples we were expecting in to our HQ and to keep track of all
+the steps that are required to process each sample: data attribution, fit
+analysis, photography, videography etc. Again this project would require loading
+multiple rows of data built from many related Postgres tables, only this time
+searching, sorting and the ability to see many days in a single view were much
+more important features.
+
+We had been using [ElasticSearch][elasticsearch] for a couple of years and never really taken
+advantage of it as a data store (as opposed to using it as a search engine and
+loading ActiveRecord objects found from the primary key ids it returned). The
+advantages were obvious: ElasticSearch indexes are very easy to search and 
+with no ActiveRecord objects to instantiate and no need to hit the database the
+response time for even large numbers of records would be acceptable.
+
+[elasticsearch]: http://www.elasticsearch.org/
+
+All we needed to do was build a search index that contained all the data we
+wanted to display in Sample Queue view layer. However, we realised it would be
+useful if we could swap out individual rows of the Sample Queue with
+ActiveRecord objects or JSON hashes. Or, to put it another way, the view layer
+shouldn't care about the class of the object it was rendering.
+
+### Presenters as Denormalizers
+The [Presenter pattern][presenter-pattern] is a familiar one in Rails: wrap
+objects in a presenter class to keep business logic away from your HAML or ERB.
+Since the process of building up a hash to push into an ElasticSearch index is basically the same
+we decided to use one class to do both jobs. The `SampleDenormalizer` looks like
+this:
+
+[presenter-pattern]: http://nithinbekal.com/posts/rails-presenters/ "Nithin Bekal on Presenters in Rails"
+
+```ruby
+class SampleDenormalizer
+
+  attr_reader :sample
```

Comment:
Does this need to be here?  Other than Dave's-particular-rules-of-coding, if it's not needed to make the point in the article, best to remove it.